core/src/test/java/io/questdb/griffin/KeyedAggregationTest.java

```diff
@@ -26,20 +26,251 @@
+            }
+
+            @Override
+            public boolean reset(long pRosti, int toSize) {
+                Rosti.disableOOMOnMalloc();
+                return super.reset(pRosti, toSize);
+            }
+        };
+
+        executeWithPool(WORKER_COUNT, 64, raf, (CairoEngine engine, SqlCompiler compiler, SqlExecutionContext sqlExecutionContext) -> {
+            compiler.compile("create table tab as " +
+                    "(select rnd_double() d, x, " +
+                    "rnd_int() i, " +
+                    "rnd_long() l, " +
+                    "rnd_date(to_date('2015', 'yyyy'), to_date('2022', 'yyyy'), 0) dat, " +
+                    "rnd_timestamp(to_timestamp('2015','yyyy'),to_timestamp('2022','yyyy'),0) tstmp " +
+                    "from long_sequence(100))", sqlExecutionContext);
+            compiler.compile("alter table tab add column s symbol cache", sqlExecutionContext).execute(null).await();
+            compiler.compile("insert into tab select rnd_double(), " +
+                    "x + 1000,   " +
+                    "rnd_int() i, " +
+                    "rnd_long() l, " +
+                    "rnd_date(to_date('2015', 'yyyy'), to_date('2022', 'yyyy'), 0) dat, " +
+                    "rnd_timestamp(to_timestamp('2015','yyyy'),to_timestamp('2022','yyyy'),0) tstmp, " +
+                    "cast('s' || x as symbol)  from long_sequence(896)", sqlExecutionContext);
+
+            final String[] functions = {"sum(d)", "min(d)", "max(d)", "avg(d)", "nsum(d)", "ksum(d)",
+                    "sum(tstmp)", "min(tstmp)", "max(tstmp)",
+                    "sum(l)", "min(l)", "max(l)", "avg(l)",
+                    "sum(i)", "min(i)", "min(i)", "avg(i)",
+                    "sum(dat)", "min(dat)", "max(dat)"};
+
+            for (String function : functions) {
+                String query = "select s, " + function + " from tab group by s";
+                LOG.infoW().$(function).$();
+                //String query = "select s, sum(d) from tab group by s";
+                assertRostiMemory(compiler, query, sqlExecutionContext);
+                sizeCounter.set(0);
+            }
+        });
+    }
+
+    private void assertRostiMemory(SqlCompiler compiler, String query, SqlExecutionContext sqlExecutionContext) throws SqlException {
+        long memBefore = Unsafe.getMemUsedByTag(MemoryTag.NATIVE_ROSTI);
+        try (final RecordCursorFactory factory = compiler.compile(query, sqlExecutionContext).getRecordCursorFactory()) {
+            try {
+                try (RecordCursor cursor = factory.getCursor(sqlExecutionContext)) {
+                    cursor.hasNext();
+                }
+                Assert.fail("Cursor should throw OOM for " + query + " !");
+            } catch (OutOfMemoryError oome) {
+                long memAfter = Unsafe.getMemUsedByTag(MemoryTag.NATIVE_ROSTI);
+                MatcherAssert.assertThat(memAfter, is(lessThan(memBefore + 10 * 1024)));//rostis are minimized on OOM
```

Comment:
10* 1024 ???

Reply:
Yes,  rostis aren't deallocated completely but minimized because even if there's OOM it doesn't necessarily mean that factory will be closed and dropped . It could still remain in cache and be used again later  (e.g. with different bind variable values ).